### PR TITLE
Not sure if this template refactor can support all test scenario in the future

### PR DESCRIPTION
**Pull request by tangzhankun**
I tried to direct mock the original virtual methods of MesosSchedulerDriver using class Inheritance, cmpiled passed but failed to new the MockMesosSchedulerDriver when running. It will hung there forever. I don't know the reason. So I choose mock non-virtual methods with template.

1. use template to refactor CephSchedulerAgent to support gmock MesosSchedulerDriver;

2. write a simple test case for testing role resource check

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -107,7 +107,6 @@ set(CEPH_MESOS_SCHEDULER_SOURCES
     src/httpserver/FileServer.cpp
     src/httpserver/RestServer.cpp
     src/scheduler/CephScheduler.cpp
-    src/scheduler/CephSchedulerAgent.cpp
```
Comment (zyluo) [inline CMakeLists.txt]: Is this intended?

File: CMakeLists.txt [hunk 2]
```diff
@@ -121,6 +120,13 @@ set(CEPH_MESOS_TEST_SOURCES
     src/tests/main.cpp
     src/tests/ConfigTest.cpp
     src/common/Config.cpp
+
```
Comment (zyluo) [inline CMakeLists.txt]: please create another set of test sources

set(CEPH_MESOS_SCHEDULER_TEST_SOURCES
...
)

File: CMakeLists.txt (hunk 1 above)
Comment (tangzhankun) [inline CMakeLists.txt]: yes. template implementation don't need this cpp. All in one .hpp file

File: CMakeLists.txt [hunk 3]
```diff
@@ -164,5 +170,10 @@ target_link_libraries(ceph-mesos-tests
     ${PTHREAD_LIB}
     ${GFLAGS_LIB}
     ${YAMLCPP_LIB}
+    ${MICROHTTPD_LIB}
+    ${MESOS_LIB}
+    ${JSONCPP_LIB}
+    ${PTHREAD_LIB}
+    ${RT_LIB}
 )
 add_test(config-tests ceph-mesos-tests)
```
Comment (zyluo) [inline CMakeLists.txt]: and

add_executable(ceph-mesos-scheduler-tests $CEPH_MESOS_SCHEDULER_TEST_SOURCES)
target_include_directories(ceph-mesos-scheduler-tests
...
)
add_test(scheduler-tests ceph-mesos-scheduler-tests)

File: CMakeLists.txt (hunk 1 above)
Comment (zyluo) [inline CMakeLists.txt]: Oh so you changed to using templates

Comment (tangzhankun) [inline CMakeLists.txt]: yes. If we don't use template, I cannot mock MesosSchedulerDriver. The ideal way is to inherit the MesosSchedulerDriver when mock it, but it failed to new instance when running test case.

File: CMakeLists.txt (hunk 3 above)
Comment (tangzhankun) [inline CMakeLists.txt]: would it be better to generate one test binary to ensure all test case passed? I am not sure

Comment (tangzhankun) [inline CMakeLists.txt]: hi, I have changed the CMakeLists.txt. please check

File: CMakeLists.txt [hunk 4]
```diff
@@ -159,10 +166,26 @@ add_executable(ceph-mesos-tests ${CEPH_MESOS_TEST_SOURCES})
...
     ${GTEST_LIB}
     ${PTHREAD_LIB}
     ${GFLAGS_LIB}
     ${YAMLCPP_LIB}
 )
+target_link_libraries(ceph-mesos-scheduler-tests
+    ${GTEST_LIB}
+    ${GFLAGS_LIB}
+    ${YAMLCPP_LIB}
+    ${MESOS_LIB}
```
Comment (zyluo) [inline CMakeLists.txt]: It would be best not to include this but let's remove this in the future. Unittests should run anywhere with minimal dependency.

Comment (tangzhankun) [inline CMakeLists.txt]: if we don't have ${mesos_lib}, compile will fail in linking stage.

Comment (zyluo) [inline CMakeLists.txt]: Is there by any chance to mock the mesos related modules?

**Pull request merged**